cmd/kube-apiserver/app/aggregator.go

```diff
@@ -102,11 +103,24 @@ func createAggregatorServer(aggregatorConfig *aggregatorapiserver.Config, delega
 		autoRegistrationController)
 
 	aggregatorServer.GenericAPIServer.AddPostStartHook("kube-apiserver-autoregistration", func(context genericapiserver.PostStartHookContext) error {
-		go autoRegistrationController.Run(5, context.StopCh)
 		go crdRegistrationController.Run(5, context.StopCh)
+		go func() {
+			// let the CRD controller process the initial set of CRDs before starting the autoregistration controller.
+			// this prevents the autoregistration controller's initial sync from deleting APIServices for CRDs that still exist.
+			crdRegistrationController.WaitForInitialSync()
+			autoRegistrationController.Run(5, context.StopCh)
+		}()
 		return nil
 	})
+
+	allAvailable := &atomic.Value{}
+	allAvailable.Store(false)
 	aggregatorServer.GenericAPIServer.AddHealthzChecks(healthz.NamedCheck("autoregister-completion", func(r *http.Request) error {
+		// Short circuit once we observed all our local services registered and available
+		if allAvailable.Load().(bool) {
+			return nil
+		}
+
 		items, err := aggregatorServer.APIRegistrationInformers.Apiregistration().InternalVersion().APIServices().Lister().List(labels.Everything())
```

Comment:
We only want to wait until everything has been observed at least once, right?  Seems like that requires a watch from a listed resource version, not a lister since this lister is vulnerable to a create, then delete.

This pull makes it much less likely if you want to fix it later.

Reply:
switched to a watch